packages/block-editor/src/store/selectors.js

```diff
@@ -192,13 +230,17 @@ export const __unstableGetBlockWithoutInnerBlocks = createSelector(
  *
  * @param {Object}  state        Editor state.
  * @param {?string} rootClientId Optional root client ID of block list.
+ * @param {?boolean} withControlledInnerBlocks If true, include inner blocks that
+ *                                             are controlled by their parent.
  *
  * @return {Object[]} Post blocks.
  */
 export const getBlocks = createSelector(
-	( state, rootClientId ) => {
+	( state, rootClientId, withControlledInnerBlocks = true ) => {
```

Comment:
Can you clarify in which situations we need this to be passed as false `withControlledInnerBlocks`?

Reply:
When getting the new blocks for comparison in the block editor provider's listener to decide if it should update the editor store or not.

Likewise, for controlled inner blocks to decide if it should update its entity or not as they may also have nested controlled inner blocks.

Reply:
Oh sorry I misread the name, when do we need it to be "true". I feel always set to "false" is what we want in most/all cases?

Reply:
Yeah, I think we could clean this up, and that would probably "fix" nested controlled inner blocks. The only struggle is the cache stuff. Currently:

- `BlockEditorProvider` calls `getBlocks` with `false`, which means that no inner blocks are included.
- `InnerBlocks` calls normal `getBlock`, which means that `getBlocks` is called with `true` for all of the inner blocks. Practically, we need this so that `InnerBlocks` sees changes to its direct innerBlocks children. 

However, I think we could default `getBlocks` to not ever include controlled inner blocks as long as the `getBlock` selector has a way to include inner blocks for the template block.

So we would basically need `getBlock( includeTopLevelInnerBlocks )`, which would basically bypass the check for controlled inner blocks _only_ for the top-level inner blocks. This way, the template part sees changes to its children.

We need that because `getBlocksWithoutControlledInnerBlocks` always checks the current client ID to see if it's a controller. If it is a controller, then it will never include controlled inner blocks, so a template part would never see its inner blocks changing.

I hope that helps explain the current situation :)

Reply:
Let me know what you think of this change: https://github.com/WordPress/gutenberg/pull/21368/files#diff-8da66aa20f6d41ab2e24dbd603db6111R152-R189

Instead of modifying `getBlocks` and adding another `getBlock` selector, I only changed the `getBlock` selector so that it excludes controlled inner blocks by default. It also has an option to include the top-level inner blocks of an inner block controller if that controller specifically asks for them. 

So in InnerBlocks, we do `getBlock( withImmediateChildren = true )`, and then the immediate controlled inner blocks are returned, but any nested controlled inner blocks are excluded. So that _should_ mean that we now support nested inner block controllers.